### ISSUE TITLE
Allow conversions to `std::string_view` etc.

### ISSUE THREAD

**Issue by alexolog** [jtv/libpqxx#694]
<Note from jtv:>

This ticket has evolved a lot.  I believe the original specific problem is solved, but more generally, I'm planning to allow conversions to view types where lifetime becomes the caller's problem.

**Comment by jtv** [comment]
Thanks for finding that.  I think I'll have to look into some ways to reduce the redundancy between those places in the code.

**Comment by alexolog** [comment]
As long as pqxx::row holds the result internally

**Comment by jtv** [comment]
> As long as pqxx::row holds the result internally

It's `pqxx::result` that holds the data, and `pqxx::row` acts as a reference to that, but yes.  In fact `pqxx::result` is basically a `std::shared_ptr` so you can copy it around and it will still refer to the same buffer.  I'm fairly sure I documented it _somewhere,_ but I'd have to make that clearer.

**Comment by alexolog** [comment]
I assume that `pqxx::row` holds a `shared_ptr` copy to the corresponding `pqxx::result`, so I can safely do something like (please ignore the syntax):
```
auto f() {
  return query.exec()[0];
}
```

**Comment by jtv** [comment]
Yes and no, respectively.

Yes, the `pqxx::row` is a reference to the `pqxx::result` which in turn is basically a `std::shared_ptr` to the data.

But in this `return` example, the `pqxx::result` itself goes out of scope when you leave the function!  And there's no other `pqxx::result` pointing to the same buffer, so that's the end of the buffer's lifetime.

**Comment by alexolog** [comment]
So `pqxx::row` does not actually hold an instance of the shared pointer itself (otherwise the memory would stick around) and is just a simple reference to an inter data structure.

I haven't checked the documentation very closely, but it would be a good idea to warn about such usage because it looks like the row is returned by value.

**Comment by jtv** [comment]
I do still intend to fix this one...  But it's a pretty radical change to allow general conversation to `string_view`, so probably best for 8.0.

**Comment by jlittlenz** [comment]
I just tripped on something related to this.  pqxx::transaction_base::query<> allows std::string_view, but not std::optional<std::string_view>.

**Comment by jtv** [comment]
@jlittlenz converting to `string_view` is a bit awkward in the general case, for lifetime reasons: it's relatively easy to get yourself into a situation where the underlying string data gets deallocated before you're done reading it.

However in libpqxx 8.0 I'm thinking to make this the caller's problem.  You would be able to "convert" any string to a `string_view`, but you would be responsible for ensuring that you don't access the string beyond its lifetime.  That does mean that I'll have to document lifetime information in a lot of places.

**Comment by jlittlenz** [comment]
Thank you.  My "trip" was partly due to my expectations of the field::view method; null values were tripping everything, but not field::view.  It just gives an empty string, and one can use field::is_null if you care about it.

**Comment by jtv** [comment]
Right, @jlittlenz.  Generally you should check for nullness before trying to read a value.

Although I did just notice that nowadays you are guaranteed to get an empty string when the value is null.  That was not the case when I first wrote libpqxx, but it is the case in all supported versions no.  So it may be time to review the design for that.  Converting a null to any type is unsupported, but for some types it may be reasonable to accept an empty string as a null.  Think `std::optional<int>` for instance.

**Comment by jtv** [comment]
Yes, this has been on the shelf for a long time...  Life has distracted me to some degree.

**Comment by jtv** [comment]
That's mildly disappointing: I can't implement a conversion from string to `pqxx::zview` because `pqxx::string_traits::from_string()` takes a `std::string_view` — and there's no way to guarantee that that value would be valid as a `pqxx::zview`.

**Comment by jtv** [comment]
The feature has been merged into the 8.0 working branch.
